### ISSUE TITLE
Launching cordova application from push notification

### ISSUE THREAD

**Issue by stt** [ExactTarget/JB4A-SDK-CordovaPlugin#6]
(A separate issue from #3).

What works:
When cordova application has successfully registered to ExactTarget, ETPush.registerForNotifications has been called, a push notification is delivered and when notification is clicked cordova application on background is activated.

What doesn't:
When application is not running the push notifications no longer go through and adb logcat gives the following log:

I/ActivityManager(  936): Start proc com... for broadcast com.../com.exacttarget.etpushsdk.ET_GenericReceiver: pid=24140 uid=10197 gids={50197, 9997, 3003, 1028, 1015} abi=armeabi-v7a
I/art     (24140): Late-enabling -Xcheck:jni
D/QcrilMsgTunnelSocket( 4811): readRilMessage: Buffer = [B@39f2b8d7 HexData = [010000000404000011000000514f454d484f4f4bef0308000100000001]
D/QcrilMsgTunnelSocket( 4811): Rcvd UNSOL response with 28 bytes data for SUB0
D/QcrilMsgTunnelSocket( 4811): Response ID 525295 is not served in this process.
D/QcrilMsgTunnelSocket( 4811): To broadcast an Intent via the notifier to external apps
D/QcrilMsgTunnelIfaceManager( 4811): handleMessage what = 0
D/QcrilMsgTunnelIfaceManager( 4811): Broadcasting intent ACTION_UNSOL_RESPONSE_OEM_HOOK_RAW
E/etpushsdk@ET_GenericReceiver(24140): No ETPush instance available for use. Did you forget to call readyAimFire() first?
E/etpushsdk@ET_GenericReceiver(24140): com.exacttarget.etpushsdk.ETException: No ETPush instance available for use. Did you forget to call readyAimFire() first?
E/etpushsdk@ET_GenericReceiver(24140):  at com.exacttarget.etpushsdk.ETPush.pushManager(ETPush.java:130)
E/etpushsdk@ET_GenericReceiver(24140):  at com.exacttarget.etpushsdk.ET_GenericReceiver.onReceive(ET_GenericReceiver.java:159)
E/etpushsdk@ET_GenericReceiver(24140):  at android.app.ActivityThread.handleReceiver(ActivityThread.java:2608)
etc..

So the notification processing stops after GCM event is received by ET_GenericReceiver from etsdk-3.3.0.jar and it calls ETPush.pushManager() without any initialization.

Had a glance at the decompiled etsdk-3.3, 3.5 and 4.0, and before 4.0 there's a setupLaunchIntent method in ET_GenericReceiver that would call ETPushNotificationRecipient from the cordova plugin, but ETPush.pushManager gets called first, so the cordova plugin doesn't have a chance to do any initialization.
It seems this usecase of launching stopped apps for push notifications wasn't supported before 4.0 which finally tries to do initialization within ETPush.pushManager (or getInstance as it's now called).

Viable workaround is probably to declare a custom receiver class for com.google.android.c2dm.permission.SEND in AndroidManifest that wraps ET_GenericReceiver.

**Comment by androidfu** [comment]
Please post your question to http://salesforce.stackexchange.com/

**Comment by androidfu** [comment]
readyAimFire() must be called from a Class that Extends Android Application.  Calling readyAimFire() with Application Context is not enough as readyAimFire() must be called to bootstrap the SDK when Push Messages are received and your application is not running.

**Comment by stt** [comment]
I'd like to know where this is documented, because it does not appear here
 https://github.com/ExactTarget/JB4A-SDK-CordovaPlugin#android-installation
or here
 https://code.exacttarget.com/apis-sdks/journey-builder-for-apps/additional-resources/cordova.html

So if nothing else this is a bug for documentation.

(I suppose this is a similar situation with another push notification library http://stackoverflow.com/a/26367952 https://parse.com/questions/android-push-notification-error-msg)

**Comment by stt** [comment]
Not sure what you mean by "Calling readyAimFire() with Application Context is not enough", is there anything else that exists during initialization of class extending android.app.Application?

Tried creating an application class (which is not normally done for Cordova apps, which is exactly why this is pretty crucial information to document) but it doesn't seem to be called, I did declare it in AndroidManifest.xml <application> tag.
Anyway, tested the workaround of initializing ETPush in custom push receiver and it seems to work fine for me?

	public class Initing_ET_GenericReceiver extends BroadcastReceiver {
	  public final void onReceive(Context context, Intent intent) {
	      ETSdkWrapper etsdk = new ETSdkWrapper();
	      etsdk.context = context;
	      etsdk.initApp();
	      new ET_GenericReceiver().onReceive(context, intent);
	  }
	}

I still consider this an issue with JB4A-SDK-CordovaPlugin and not a support request, even if it's just about the documentation (if you only knew how much wasted time a small example could have saved).
